### PR TITLE
Make Quadrotor C++ slightly more idiomatic

### PR DESCRIPTION
**Pull request by jwnimmer-tri**
I stumbled upon this when investigating transmogrification.

---
This change is [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/6620)

### CODE REVIEW

**Comment by jwnimmer-tri** [comment]
+@naveenoid for feature review, please.

**Comment by naveenoid** [comment]
:lgtm:

---

Review status: 0 of 1 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6620#-:-Kp_IycuBBatVmbCF6lS:bnfp4nl)*

**Comment by jwnimmer-tri** [comment]
+@SeanCurtis-TRI for platform review, please.

**Comment by SeanCurtis-TRI** [comment]
:LGTM: with a weak request. +@sherm1 for final platform review.

---

Reviewed 1 of 1 files at r1.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/examples/Quadrotor/quadrotor_plant.h, line 62 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6620#-Kp_gVxGh7W35g8pehKL:-Kp_gVxGh7W35g8pehKM:b-3yriie) ([raw file](https://github.com/robotlocomotion/drake/blob/9afef378b33fead390dc6d0cecadbe8b5dda9adb/drake/examples/Quadrotor/quadrotor_plant.h#L62)):*
> ```objc
>  private:
>   // TODO(naveenoid): Declare these as parameters in the context.
>   const double g_{9.81};     // Gravitational acceleration (m/s^2).
> ```

BTW We've got some `const` member naming inconsistency.
1. Two of six are missing the `k` prefix.
2. Two of six do not have a trailing `_`.

I realize this PR is about addressing C++ idiom, but this touches some style issues. Correct both, perhaps?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6620#-:-Kp_heM0zrKe3hU6Gztx:brvt8p8)*

**Comment by jwnimmer-tri** [comment]
*[drake/examples/Quadrotor/quadrotor_plant.h, line 62 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6620#-Kp_gVxGh7W35g8pehKL:-Kp_hyLNtwDJgKDCXgOm:bxid99f) ([raw file](https://github.com/robotlocomotion/drake/blob/9afef378b33fead390dc6d0cecadbe8b5dda9adb/drake/examples/Quadrotor/quadrotor_plant.h#L62)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

BTW We've got some `const` member naming inconsistency.
1. Two of six are missing the `k` prefix.
2. Two of six do not have a trailing `_`.

I realize this PR is about addressing C++ idiom, but this touches some style issues. Correct both, perhaps?
</blockquote></details>

The lack of `_` is a defect that I'd missed.  I'll fix it.  (Good catch.)

The "kF" and "kM" here are not actually as in compile-time "kConstant" from cppguide (you'll note that the second constructor lets a user change them, which means they are not kConstant for the duration of the program) -- as I understand it, the algebraic variable name from the model is actually a "k".  I'd propose leaving them alone, because they should really be `systems::Parameters` so this code should die anyway.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6620)*

**Comment by SeanCurtis-TRI** [comment]
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/examples/Quadrotor/quadrotor_plant.h, line 62 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6620#-Kp_gVxGh7W35g8pehKL:-Kp_jjepXii_k5aTd4Jk:b4syurz) ([raw file](https://github.com/robotlocomotion/drake/blob/9afef378b33fead390dc6d0cecadbe8b5dda9adb/drake/examples/Quadrotor/quadrotor_plant.h#L62)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

The lack of `_` is a defect that I'd missed.  I'll fix it.  (Good catch.)

The "kF" and "kM" here are not actually as in compile-time "kConstant" from cppguide (you'll note that the second constructor lets a user change them, which means they are not kConstant for the duration of the program) -- as I understand it, the algebraic variable name from the model is actually a "k".  I'd propose leaving them alone, because they should really be `systems::Parameters` so this code should die anyway.
</blockquote></details>

A few of things:

1. Excellent point about program-life.
2. I just realized, that there are *seven* and not *six*.  The orphan is `g_`, which *should* be `kG_` - it is program-life const.
3. I'm certainly content to see these persist until they get moved into parameters. Can we get an issue assigned to...say... @naveenoid on that?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6620)*

**Comment by jwnimmer-tri** [comment]
Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/examples/Quadrotor/quadrotor_plant.h, line 62 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6620#-Kp_gVxGh7W35g8pehKL:-Kp_l8r8ditqMZP85EUq:b-o3i3jq) ([raw file](https://github.com/robotlocomotion/drake/blob/9afef378b33fead390dc6d0cecadbe8b5dda9adb/drake/examples/Quadrotor/quadrotor_plant.h#L62)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

A few of things:

1. Excellent point about program-life.
2. I just realized, that there are *seven* and not *six*.  The orphan is `g_`, which *should* be `kG_` - it is program-life const.
3. I'm certainly content to see these persist until they get moved into parameters. Can we get an issue assigned to...say... @naveenoid on that?
</blockquote></details>

Done (on the missing `_`).

Re (2.) -- if you look carefully at https://google.github.io/styleguide/cppguide.html#Constant_Names the rule is thus:
- If  you name as kConstant, then you must always have the same value;
- If you always have the same value _and are global_, then you must name as kConstant.

In this case, g_ is not global so is (by cppguide) permitted to omit the k.  Beyond cppguide, it could still be defective by being unusual or confusing, but I think in this case `g_` might be intended to turn into a `system::Parameters` later, so is plausibly justified as being non-k.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6620)*

**Comment by SeanCurtis-TRI** [comment]
Thanks

---

Reviewed 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/examples/Quadrotor/quadrotor_plant.h, line 62 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6620#-Kp_gVxGh7W35g8pehKL:-Kp_ptdRz12r9_o1O4pQ:b-13hul2) ([raw file](https://github.com/robotlocomotion/drake/blob/9afef378b33fead390dc6d0cecadbe8b5dda9adb/drake/examples/Quadrotor/quadrotor_plant.h#L62)):*
tldr; I concur.

Rant: The language on this style guide point is poor. Consider the initial sentence (emphasis mine):

> Variables declared constexpr or const, and whose value is fixed for the duration of the program, __*are*__ named with a leading "k" followed by mixed case.

It's a very strong, unambiguous statement. And then the very next sentence, in some sense, contradicts it.

> All such variables with static storage duration (i.e. statics and globals, see Storage Duration for details) should be named this way.

So, in fact, the example they give *depends* on the scope in which that statement appears.

Blech.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6620#-:-Kp_qqNthvYs--UJV44T:b-zcjvte)*

**Comment by naveenoid** [comment]
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/examples/Quadrotor/quadrotor_plant.h, line 62 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6620#-Kp_gVxGh7W35g8pehKL:-Kp_qw_DTxulNKjJ0mN_:b-z6te9a) ([raw file](https://github.com/robotlocomotion/drake/blob/9afef378b33fead390dc6d0cecadbe8b5dda9adb/drake/examples/Quadrotor/quadrotor_plant.h#L62)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

tldr; I concur.

Rant: The language on this style guide point is poor. Consider the initial sentence (emphasis mine):

> Variables declared constexpr or const, and whose value is fixed for the duration of the program, __*are*__ named with a leading "k" followed by mixed case.

It's a very strong, unambiguous statement. And then the very next sentence, in some sense, contradicts it.

> All such variables with static storage duration (i.e. statics and globals, see Storage Duration for details) should be named this way.

So, in fact, the example they give *depends* on the scope in which that statement appears.

Blech.
</blockquote></details>

Im ok with issue or even just a TODO assigned in my name for fixing this up in the future.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6620)*

**Comment by jwnimmer-tri** [comment]
*[drake/examples/Quadrotor/quadrotor_plant.h, line 62 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6620#-Kp_gVxGh7W35g8pehKL:-Kp_rTXtffeClakhjzXR:beg307i) ([raw file](https://github.com/robotlocomotion/drake/blob/9afef378b33fead390dc6d0cecadbe8b5dda9adb/drake/examples/Quadrotor/quadrotor_plant.h#L62)):*
<details><summary><i>Previously, naveenoid (Naveen Kuppuswamy) wrote…</i></summary><blockquote>

Im ok with issue or even just a TODO assigned in my name for fixing this up in the future.
</blockquote></details>

Yeah.  The wording in that part of the guide (and many other parts!) doesn't help its comprehensibility.

The "Parameters" is already captured in a TODO, one line higher up in the file than this topic.  If either Naveen or Sean also want a github issue, you can file that as you like.  I'm comfortable with just the TODO in the code.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6620)*

**Comment by sherm1** [comment]
:lgtm: with one suggestion below.

---

Reviewed 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/examples/Quadrotor/quadrotor_plant.cc, line 107 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6620#-KpaO5Z2syve2KSX3GVz:-KpaO5Z2syve2KSX3GW-:b5tkutj) ([raw file](https://github.com/robotlocomotion/drake/blob/6260840c1f08868e6eefd0b58e6d8d4cb776f2f0/drake/examples/Quadrotor/quadrotor_plant.cc#L107)):*
> ```cpp
> constexpr int QuadrotorPlant<T>::kStateDimension;
> template <typename T>
> constexpr int QuadrotorPlant<T>::kInputDimension;
> ```

FYI, providing out-of-class definitions for static constexpr members is not required and is apparently deprecated in C++17. See [this link](http://en.cppreference.com/w/cpp/language/static) (scroll down to "Constant static members").

This is also true for static const members (unless they are "odr-used" which I understand to mean someone takes their address). C++17 adds "static inline" members to make this even more contorted.

I would suggest removing these definitions.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6620#-:-KpaQCxK4NNAmLhorDIy:b-yel8ks)*

**Comment by jwnimmer-tri** [comment]
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/examples/Quadrotor/quadrotor_plant.cc, line 107 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6620#-KpaO5Z2syve2KSX3GVz:-KpaY5vlM42oixeInU3e:b-9hbm1h) ([raw file](https://github.com/robotlocomotion/drake/blob/6260840c1f08868e6eefd0b58e6d8d4cb776f2f0/drake/examples/Quadrotor/quadrotor_plant.cc#L107)):*
> providing out-of-class definitions for static constexpr members is not required

I don't think thats true?  Setting aside C++17 rules (since our compilers don't use it), the C++11 rules cited on that page seem to be:

> > "If a ... constexpr static data member is odr-used, a definition at namespace scope is still required."

To me, that says that a definition is _sometimes_ required.

Because it is difficult to know whether a constant will be odr-used, and there is no diagnostic required if it later it becomes odr-used but lack a definition, it seems sensible to me to keep these definitions intact.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6620)*

**Comment by sherm1** [comment]
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/examples/Quadrotor/quadrotor_plant.cc, line 107 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6620#-KpaO5Z2syve2KSX3GVz:-KpakYcgKEBsrFePAEHF:bafhf6a) ([raw file](https://github.com/robotlocomotion/drake/blob/6260840c1f08868e6eefd0b58e6d8d4cb776f2f0/drake/examples/Quadrotor/quadrotor_plant.cc#L107)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

> providing out-of-class definitions for static constexpr members is not required

I don't think thats true?  Setting aside C++17 rules (since our compilers don't use it), the C++11 rules cited on that page seem to be:

> > "If a ... constexpr static data member is odr-used, a definition at namespace scope is still required."

To me, that says that a definition is _sometimes_ required.

Because it is difficult to know whether a constant will be odr-used, and there is no diagnostic required if it later it becomes odr-used but lack a definition, it seems sensible to me to keep these definitions intact.
</blockquote></details>

A *private* static member can't become odr-used unless the class takes its address locally -- that isn't happening now and seems very unlikely! There would certainly be a link-time diagnostic if the definition is needed but unavailable. And this will have to be removed when we switch to C++17 (if I'm understanding that page correctly). So I wouldn't add these definitions, but of course you can if you prefer!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6620)*
